stageleft_macro/src/lib.rs

```diff
@@ -0,0 +1,365 @@
+                                __quse_local
+                            }
+                        }
+                    );
+                });
+            } else if !prefix.is_empty() {
+                let first = prefix.first().unwrap();
+                let prefix_suffix = prefix.iter().skip(1);
+                let suffix_full_path = quote!(#(#prefix_suffix::)*#name_ident).to_string();
+
+                into.push(quote! {
+                    #[allow(non_upper_case_globals, non_snake_case)]
+                    let #name_ident_unspanned = #first.extend(
+                        #suffix_full_path,
+                        {
+                            let __quse_local = ();
+                            {
+                                use #(#prefix_unspanned::)*#name_ident_unspanned as __quse_local;
+                                __quse_local
+                            }
+                        }
+                    );
+                });
+            } else {
+                into.push(quote! {
+                    #[allow(non_upper_case_globals, non_snake_case)]
+                    let #name_ident = #root::runtime_support::Import::clone(&#name_ident);
+                });
+            }
+        }
+        _ => todo!(),
+    }
+}
+
+#[proc_macro]
+pub fn quse_fn(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let stageleft_crate = proc_macro_crate::crate_name("stageleft")
+        .expect("stageleft should be present in `Cargo.toml`");
+    let root = match stageleft_crate {
+        proc_macro_crate::FoundCrate::Itself => quote! { stageleft },
+        proc_macro_crate::FoundCrate::Name(name) => {
+            let ident = syn::Ident::new(&name, Span::call_site());
+            quote! { #ident }
+        }
+    };
+
+    let input_tokens = proc_macro2::TokenStream::from(input);
+    let import: syn::ItemUse = syn::parse_quote!(use #input_tokens;);
+    let mut all_paths_emitted = vec![];
+    gen_use_paths(
+        root,
+        import.leading_colon.is_some(),
+        vec![],
+        &import.tree,
+        &mut all_paths_emitted,
+    );
+
+    quote! {
+        use #input_tokens;
+        #(#all_paths_emitted;)*
+    }
+    .into()
+}
+
+/// Marks a chunk of code as being runtime-only, which means that no staged code in its crate can depend on it.
+/// Code behind this attribute is allowed to use staged entrypoints defined in the same crate.
+#[proc_macro_attribute]
+pub fn runtime(
+    _attr: proc_macro::TokenStream,
+    input: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    // TODO(shadaj): when this is used for a struct, emit two versions:
+    // one that is runtime-only with public fields and one that is staged-only without
+    let input: TokenStream = input.into();
+    proc_macro::TokenStream::from(quote! {
+        #[cfg(not(feature = "macro"))]
+        #input
+    })
+}
+
+#[proc_macro_attribute]
```

Comment:
this needs documentation, I can see in places it takes some attribute arguments but I can't figure out what those mean easily.

for example:

```
#[stageleft::entry(&'static str)]
```